File: conformance/reference-js/host-conformance-claim.js
'use strict';

/**
 * Host conformance claim validator (C1).
 *
 * Implements ADR-0037 / spec 10 §4: canonical serialization and reproducible
 * digest (D1a / §4.3), capability registry validation with dependency DAG
 * (§4.4), removedIn handling, and the claim validation order of §4.8.
 * Outputs are deterministic and byte-comparable with the Python reference
 * (host_conformance_claim.py).
 */

const crypto = require('node:crypto');

const VERSION_PATTERN = /^(?:0|[1-9][0-9]*)\.(?:0|[1-9][0-9]*)$/;
// v2.8: segments may contain hyphens (host.failure-recovery) — ADR-0034 grammar widening.
const CAPABILITY_PATTERN = /^[a-z][a-z0-9-]*(?:\.[a-z][a-z0-9-]*)*$/;
const SUITE_ID_PATTERN = /^[a-z0-9-]+$/;

function isObject(value) {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

/** Numeric MAJOR.MINOR comparison (null-safe: null versions compare as absent). */
function compareVersion(left, right) {
  const [leftMajor, leftMinor] = left.split('.').map(Number);
  const [rightMajor, rightMinor] = right.split('.').map(Number);
  if (leftMajor !== rightMajor) return leftMajor < rightMajor ? -1 : 1;
  if (leftMinor !== rightMinor) return leftMinor < rightMinor ? -1 : 1;
  return 0;
}

/** Array of strings sorted by UTF-8 byte order (=== codepoint order). */
function isStringArray(value) {
  return Array.isArray(value) && value.every(item => typeof item === 'string');
}

function isObjectArray(value) {
  return Array.isArray(value) && value.every(item => isObject(item));
}

/**
 * Canonical serialization (D1a / §4.3): object keys byte-ascending, string
 * arrays byte-ascending, object arrays sorted by canonical bytes, minimal
 * RFC 8259 escaping, no whitespace.
 *
 * Number support is restricted to integers: valid claims contain no numeric
 * fields (C0), and JS/Python float exponent formatting differs
 * (1e-7 vs 1e-07), which would break cross-language digest equality.
 */
function canonicalize(value) {
  if (isObject(value)) {
    const keys = Object.keys(value).sort();
    const parts = [];
    for (const key of keys) {
      parts.push(`${JSON.stringify(key)}:${canonicalize(value[key])}`);
    }
    return `{${parts.join(',')}}`;
  }
  if (Array.isArray(value)) {
    let items = value;
    if (isStringArray(items)) {
      items = [...items].sort();
    } else if (isObjectArray(items)) {
      items = [...items].sort((left, right) => {
        const leftBytes = canonicalize(left);
        const rightBytes = canonicalize(right);
        return leftBytes < rightBytes ? -1 : leftBytes > rightBytes ? 1 : 0;
      });
    }
    return `[${items.map(item => canonicalize(item)).join(',')}]`;
  }
  return JSON.stringify(value);
}

/** SHA-256 (lowercase hex) over the canonical UTF-8 bytes. */
function claimDigest(claim) {
  return crypto.createHash('sha256').update(canonicalize(claim), 'utf8').digest('hex');
}

/**
 * Registry validation (§4.4): closed entries, precise MAJOR.MINOR or null
 * versions, removedIn strictly after deprecatedSince, existing dependency
 * targets, unique IDs and an acyclic dependency graph.
 */
function validateRegistry(registry) {
  const errors = [];
  if (!isObject(registry) || !isObject(registry.capabilities)) {
    return { valid: false, errors: ['registry.capabilities must be an object'] };
  }
  const capabilities = registry.capabilities;
  const ids = Object.keys(capabilities);

  for (const id of ids) {
    if (!CAPABILITY_PATTERN.test(id)) {
      errors.push(`invalid capabilityId: ${id}`);
      continue;
    }
    const entry = capabilities[id];
    if (!isObject(entry)) {
      errors.push(`${id}: entry must be an object`);
      continue;
    }
    const allowedKeys = new Set([
      'sinceProtocolVersion', 'dependsOn', 'mandatorySuites', 'deprecatedSince', 'removedIn',
    ]);
    for (const key of Object.keys(entry)) {
      if (!allowedKeys.has(key)) errors.push(`${id}: unknown key ${key}`);
    }
    for (const versionKey of ['sinceProtocolVersion']) {
      if (typeof entry[versionKey] !== 'string' || !VERSION_PATTERN.test(entry[versionKey])) {
        errors.push(`${id}: ${versionKey} must be a precise MAJOR.MINOR`);
      }
    }
    for (const versionKey of ['deprecatedSince', 'removedIn']) {
      const value = entry[versionKey];
      if (value !== null && (typeof value !== 'string' || !VERSION_PATTERN.test(value))) {
        errors.push(`${id}: ${versionKey} must be MAJOR.MINOR or null`);
      }
    }
    if (entry.removedIn !== null && entry.removedIn !== undefined) {
      if (entry.deprecatedSince === null || entry.deprecatedSince === undefined) {
        errors.push(`${id}: removedIn requires deprecatedSince`);
      } else if (compareVersion(entry.removedIn, entry.deprecatedSince) <= 0) {
        errors.push(`${id}: removedIn must be strictly later than deprecatedSince`);
      }
    }
    if (!Array.isArray(entry.dependsOn) || !isStringArray(entry.dependsOn)
      || new Set(entry.dependsOn).size !== entry.dependsOn.length) {
      errors.push(`${id}: dependsOn must be a unique string array`);
    } else {
      for (const dependency of entry.dependsOn) {
        if (!Object.prototype.hasOwnProperty.call(capabilities, dependency)) {
          errors.push(`${id}: dependsOn references unregistered capability ${dependency}`);
        }
      }
    }
    if (!Array.isArray(entry.mandatorySuites) || entry.mandatorySuites.length === 0
      || !entry.mandatorySuites.every(suite => typeof suite === 'string' && SUITE_ID_PATTERN.test(suite))
      || new Set(entry.mandatorySuites).size !== entry.mandatorySuites.length) {
      errors.push(`${id}: mandatorySuites must be a non-empty unique suite-id array`);
    }
  }

  // Dependency graph must be acyclic.
  const visiting = new Set();
  const visited = new Set();
  const visit = (id, chain) => {
    if (visited.has(id)) return;
    if (visiting.has(id)) {
      errors.push(`dependency cycle detected: ${[...chain, id].join(' -> ')}`);
      return;
    }
    visiting.add(id);
    const entry = capabilities[id];
    if (isObject(entry) && Array.isArray(entry.dependsOn)) {
      for (const dependency of entry.dependsOn) {
        if (Object.prototype.hasOwnProperty.call(capabilities, dependency)) {
          visit(dependency, [...chain, id]);
        }
      }
    }
    visiting.delete(id);
    visited.add(id);
  };
  for (const id of ids) visit(id, []);

  return { valid: errors.length === 0, errors };
}

/**
 * Claim validation in the §4.8 order:
 * INVALID_CLAIM → UNKNOWN_CLAIM_VERSION → UNKNOWN_CLAIM_CAPABILITY (incl.
 * removedIn) → INCOMPLETE_CAPABILITY_DEPENDENCY → CLAIM_ARTIFACT_MISMATCH →
 * CLAIM_FIXTURE_MISMATCH → CLAIM_SUITE_INCOMPLETE → CLAIM_EVIDENCE_BUILD_MISMATCH
 * → CLAIM_EVIDENCE_UNVERIFIABLE → CLAIM_OK.
 *
 * @param {object} claim
 * @param {object} options
 * @param {object} options.registry capability-registry.json object
 * @param {string} options.artifactContentSha256 hex digest of the protocol artifact content the build consumed
 * @param {string} options.fixtureSha256 hex digest declared by the protocol artifact for the versioned fixtures
 * @param {string[]} [options.evidenceStates] per-evidence 'verifiable' | 'unverifiable' (default all verifiable)
 */
function validateClaim(claim, options = {}) {
  const registry = isObject(options.registry) && isObject(options.registry.capabilities)
    ? options.registry
    : null;
  const artifactContentSha256 = options.artifactContentSha256;
  const fixtureSha256 = options.fixtureSha256;
  const evidenceStates = Array.isArray(options.evidenceStates) ? options.evidenceStates : [];

  const invalid = (code) => ({ code });

  if (!isObject(claim)) return invalid('INVALID_CLAIM');
  if (!isObject(claim.host) || !isObject(claim.protocolArtifact)
    || !isObject(claim.support) || !isObject(claim.conformance)
    || !Array.isArray(claim.evidence)) {
    return invalid('INVALID_CLAIM');
  }
  const { pageVersions, manifestVersions, capabilities } = claim.support;
  for (const list of [pageVersions, manifestVersions]) {
    if (!Array.isArray(list) || list.length === 0 || !isStringArray(list)
      || list.some(version => !VERSION_PATTERN.test(version)) || new Set(list).size !== list.length) {
      return invalid('INVALID_CLAIM');
    }
  }
  if (!Array.isArray(capabilities) || capabilities.length === 0 || !isStringArray(capabilities)
    || capabilities.some(id => !CAPABILITY_PATTERN.test(id)) || new Set(capabilities).size !== capabilities.length) {
    return invalid('INVALID_CLAIM');
  }
  if (typeof claim.conformance.fixtureSha256 !== 'string' || !/^[0-9a-f]{64}$/.test(claim.conformance.fixtureSha256)
    || typeof claim.protocolArtifact.contentSha256 !== 'string' || !/^[0-9a-f]{64}$/.test(claim.protocolArtifact.contentSha256)) {
    return invalid('INVALID_CLAIM');
  }
  if (!Array.isArray(claim.conformance.suites) || claim.conformance.suites.length === 0) {
    return invalid('INVALID_CLAIM');
  }
  if (claim.evidence.length === 0) return invalid('INVALID_CLAIM');
  if (typeof claim.host.buildId !== 'string' || claim.host.buildId.length === 0) {
    return invalid('INVALID_CLAIM');
  }

  if (claim.claimVersion !== '1.0') return invalid('UNKNOWN_CLAIM_VERSION');
  if (!registry) return invalid('INVALID_CLAIM');

  // Capability registration + removedIn per listed version (D2 / §4.2, §4.4).
  const registryCapabilities = registry.capabilities;
  const listedVersions = [...pageVersions, ...manifestVersions];
  for (const capability of capabilities) {
    const entry = registryCapabilities[capability];
    if (!entry) return invalid('UNKNOWN_CLAIM_CAPABILITY');
    if (entry.removedIn !== null && entry.removedIn !== undefined) {
      for (const version of listedVersions) {
        if (compareVersion(entry.removedIn, version) <= 0) {
          return invalid('UNKNOWN_CLAIM_CAPABILITY');
        }
      }
    }
  }

  // Dependency closure (D2 / §4.2).
  const listed = new Set(capabilities);
  for (const capability of capabilities) {
    const pending = [...registryCapabilities[capability].dependsOn];
    while (pending.length > 0) {
      const dependency = pending.pop();
      if (!listed.has(dependency)) return invalid('INCOMPLETE_CAPABILITY_DEPENDENCY');
      const dependencyEntry = registryCapabilities[dependency];
      if (dependencyEntry && Array.isArray(dependencyEntry.dependsOn)) {
        pending.push(...dependencyEntry.dependsOn);
      }
    }
  }

  if (claim.protocolArtifact.contentSha256 !== artifactContentSha256) {
    return invalid('CLAIM_ARTIFACT_MISMATCH');
  }
  if (claim.conformance.fixtureVersion !== '1.0' || claim.conformance.fixtureSha256 !== fixtureSha256) {
    return invalid('CLAIM_FIXTURE_MISMATCH');
  }

  // Mandatory suite coverage (D4 / §4.6): union over listed capabilities.
  const suiteResults = new Map();
  for (const suite of claim.conformance.suites) {
    suiteResults.set(suite.suiteId, suite);
  }
  for (const capability of capabilities) {
    for (const suiteId of registryCapabilities[capability].mandatorySuites) {
      const suite = suiteResults.get(suiteId);
      if (!suite || suite.result !== 'pass') return invalid('CLAIM_SUITE_INCOMPLETE');
    }
  }

  // Evidence build binding (D5 / §4.7).
  for (const evidence of claim.evidence) {
    if (!isObject(evidence) || evidence.subjectBuildId !== claim.host.buildId) {
      return invalid('CLAIM_EVIDENCE_BUILD_MISMATCH');
    }
  }

  for (let index = 0; index < claim.evidence.length; index += 1) {
    const state = index < evidenceStates.length ? evidenceStates[index] : 'verifiable';
    if (state === 'unverifiable') return invalid('CLAIM_EVIDENCE_UNVERIFIABLE');
  }

  return invalid('CLAIM_OK');
}

/** Fixture dispatch facade: operation selects the semantic entry point. */
function execute(input) {
  const operation = input && input.operation;
  if (operation === 'canonicalize') return { canonical: canonicalize(input.value) };
  if (operation === 'claimDigest') return { digest: claimDigest(input.claim) };
  if (operation === 'validateRegistry') return validateRegistry(input.registry);
  if (operation === 'validateClaim') return validateClaim(input.claim, input.options);
  throw new Error(`Unknown operation: ${operation}`);
}

module.exports = { execute, canonicalize, claimDigest, validateRegistry, validateClaim };

File: conformance/reference-python/host_conformance_claim.py
"""
Host conformance claim validator (C1, Python reference).

Implements ADR-0037 / spec 10 §4. Outputs are deterministic and
byte-comparable with the JavaScript reference (host-conformance-claim.js).
"""

import hashlib
import json
import re

VERSION_PATTERN = re.compile(r"^(?:0|[1-9][0-9]*)\.(?:0|[1-9][0-9]*)$")
# v2.8: segments may contain hyphens (host.failure-recovery) — ADR-0034 grammar widening.
CAPABILITY_PATTERN = re.compile(r"^[a-z][a-z0-9-]*(?:\.[a-z][a-z0-9-]*)*$")
SUITE_ID_PATTERN = re.compile(r"^[a-z0-9-]+$")


def _is_object(value):
    return isinstance(value, dict)


def compare_version(left, right):
    """Numeric MAJOR.MINOR comparison."""
    left_major, left_minor = (int(part) for part in left.split("."))
    right_major, right_minor = (int(part) for part in right.split("."))
    if left_major != right_major:
        return -1 if left_major < right_major else 1
    if left_minor != right_minor:
        return -1 if left_minor < right_minor else 1
    return 0


def _is_string_array(value):
    return isinstance(value, list) and all(isinstance(item, str) for item in value)


def _is_object_array(value):
    return isinstance(value, list) and all(_is_object(item) for item in value)


def canonicalize(value):
    """Canonical serialization (D1a / §4.3).

    Number support is restricted to integers: valid claims contain no numeric
    fields (C0), and JS/Python float exponent formatting differs
    (1e-7 vs 1e-07), which would break cross-language digest equality.
    """
    if _is_object(value):
        parts = []
        for key in sorted(value.keys()):
            parts.append(json.dumps(key, ensure_ascii=False, separators=(",", ":"))
                         + ":" + canonicalize(value[key]))
        return "{" + ",".join(parts) + "}"
    if isinstance(value, list):
        items = value
        if _is_string_array(items):
            items = sorted(items)
        elif _is_object_array(items):
            items = sorted(items, key=lambda item: canonicalize(item))
        return "[" + ",".join(canonicalize(item) for item in items) + "]"
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def claim_digest(claim):
    """SHA-256 (lowercase hex) over the canonical UTF-8 bytes."""
    return hashlib.sha256(canonicalize(claim).encode("utf-8")).hexdigest()


def validate_registry(registry):
    """Registry validation (§4.4): closed entries, precise versions, DAG acyclic."""
    errors = []
    if not _is_object(registry) or not _is_object(registry.get("capabilities")):
        return {"valid": False, "errors": ["registry.capabilities must be an object"]}
    capabilities = registry["capabilities"]
    ids = list(capabilities.keys())

    for capability_id in ids:
        if CAPABILITY_PATTERN.fullmatch(capability_id) is None:
            errors.append("invalid capabilityId: %s" % capability_id)
            continue
        entry = capabilities[capability_id]
        if not _is_object(entry):
            errors.append("%s: entry must be an object" % capability_id)
            continue
        allowed_keys = {"sinceProtocolVersion", "dependsOn", "mandatorySuites", "deprecatedSince", "removedIn"}
        for key in entry.keys():
            if key not in allowed_keys:
                errors.append("%s: unknown key %s" % (capability_id, key))
        if not isinstance(entry.get("sinceProtocolVersion"), str) or VERSION_PATTERN.fullmatch(entry.get("sinceProtocolVersion", "")) is None:
            errors.append("%s: sinceProtocolVersion must be a precise MAJOR.MINOR" % capability_id)
        for version_key in ("deprecatedSince", "removedIn"):
            value = entry.get(version_key)
            if value is not None and (not isinstance(value, str) or VERSION_PATTERN.fullmatch(value) is None):
                errors.append("%s: %s must be MAJOR.MINOR or null" % (capability_id, version_key))
        if entry.get("removedIn") is not None:
            if entry.get("deprecatedSince") is None:
                errors.append("%s: removedIn requires deprecatedSince" % capability_id)
            elif compare_version(entry["removedIn"], entry["deprecatedSince"]) <= 0:
                errors.append("%s: removedIn must be strictly later than deprecatedSince" % capability_id)
        depends_on = entry.get("dependsOn")
        if not _is_string_array(depends_on) or len(set(depends_on)) != len(depends_on):
            errors.append("%s: dependsOn must be a unique string array" % capability_id)
        else:
            for dependency in depends_on:
                if dependency not in capabilities:
                    errors.append("%s: dependsOn references unregistered capability %s" % (capability_id, dependency))
        mandatory_suites = entry.get("mandatorySuites")
        if not isinstance(mandatory_suites, list) or len(mandatory_suites) == 0 \
                or not all(isinstance(suite, str) and SUITE_ID_PATTERN.fullmatch(suite) for suite in mandatory_suites) \
                or len(set(mandatory_suites)) != len(mandatory_suites):
            errors.append("%s: mandatorySuites must be a non-empty unique suite-id array" % capability_id)

    # Dependency graph must be acyclic.
    visiting = set()
    visited = set()

    def visit(node, chain):
        if node in visited:
            return
        if node in visiting:
            errors.append("dependency cycle detected: %s" % " -> ".join(chain + [node]))
            return
        visiting.add(node)
        entry = capabilities.get(node)
        if _is_object(entry) and isinstance(entry.get("dependsOn"), list):
            for dependency in entry.get("dependsOn", []):
                if dependency in capabilities:
                    visit(dependency, chain + [node])
        visiting.discard(node)
        visited.add(node)

    for capability_id in ids:
        visit(capability_id, [])

    return {"valid": len(errors) == 0, "errors": errors}


def validate_claim(claim, options=None):
    """Claim validation in the §4.8 order."""
    options = options if _is_object(options) else {}
    registry = options.get("registry") if _is_object(options.get("registry")) and _is_object(options.get("registry", {}).get("capabilities")) else None
    artifact_content_sha256 = options.get("artifactContentSha256")
    fixture_sha256 = options.get("fixtureSha256")
    evidence_states = options.get("evidenceStates") if isinstance(options.get("evidenceStates"), list) else []

    def invalid(code):
        return {"code": code}

    if not _is_object(claim):
        return invalid("INVALID_CLAIM")
    if not _is_object(claim.get("host")) or not _is_object(claim.get("protocolArtifact")) \
            or not _is_object(claim.get("support")) or not _is_object(claim.get("conformance")) \
            or not isinstance(claim.get("evidence"), list):
        return invalid("INVALID_CLAIM")

    support = claim["support"]
    page_versions = support.get("pageVersions")
    manifest_versions = support.get("manifestVersions")
    capabilities = support.get("capabilities")
    for version_list in (page_versions, manifest_versions):
        if not isinstance(version_list, list) or len(version_list) == 0 or not _is_string_array(version_list) \
                or any(VERSION_PATTERN.fullmatch(version) is None for version in version_list) \
                or len(set(version_list)) != len(version_list):
            return invalid("INVALID_CLAIM")
    if not isinstance(capabilities, list) or len(capabilities) == 0 or not _is_string_array(capabilities) \
            or any(CAPABILITY_PATTERN.fullmatch(capability) is None for capability in capabilities) \
            or len(set(capabilities)) != len(capabilities):
        return invalid("INVALID_CLAIM")

    conformance = claim["conformance"]
    protocol_artifact = claim["protocolArtifact"]
    sha256_pattern = re.compile(r"^[0-9a-f]{64}$")
    if not isinstance(conformance.get("fixtureSha256"), str) or sha256_pattern.fullmatch(conformance.get("fixtureSha256", "")) is None \
            or not isinstance(protocol_artifact.get("contentSha256"), str) or sha256_pattern.fullmatch(protocol_artifact.get("contentSha256", "")) is None:
        return invalid("INVALID_CLAIM")
    if not isinstance(conformance.get("suites"), list) or len(conformance.get("suites", [])) == 0:
        return invalid("INVALID_CLAIM")
    if len(claim["evidence"]) == 0:
        return invalid("INVALID_CLAIM")
    if not isinstance(claim["host"].get("buildId"), str) or len(claim["host"].get("buildId", "")) == 0:
        return invalid("INVALID_CLAIM")

    if claim.get("claimVersion") != "1.0":
        return invalid("UNKNOWN_CLAIM_VERSION")
    if registry is None:
        return invalid("INVALID_CLAIM")

    registry_capabilities = registry["capabilities"]
    listed_versions = page_versions + manifest_versions
    for capability in capabilities:
        entry = registry_capabilities.get(capability)
        if entry is None:
            return invalid("UNKNOWN_CLAIM_CAPABILITY")
        if entry.get("removedIn") is not None:
            for version in listed_versions:
                if compare_version(entry["removedIn"], version) <= 0:
                    return invalid("UNKNOWN_CLAIM_CAPABILITY")

    listed = set(capabilities)
    for capability in capabilities:
        pending = list(registry_capabilities[capability].get("dependsOn", []))
        while pending:
            dependency = pending.pop()
            if dependency not in listed:
                return invalid("INCOMPLETE_CAPABILITY_DEPENDENCY")
            dependency_entry = registry_capabilities.get(dependency)
            if dependency_entry is not None and isinstance(dependency_entry.get("dependsOn"), list):
                pending.extend(dependency_entry["dependsOn"])

    if protocol_artifact.get("contentSha256") != artifact_content_sha256:
        return invalid("CLAIM_ARTIFACT_MISMATCH")
    if conformance.get("fixtureVersion") != "1.0" or conformance.get("fixtureSha256") != fixture_sha256:
        return invalid("CLAIM_FIXTURE_MISMATCH")

    suite_results = {suite.get("suiteId"): suite for suite in conformance["suites"]}
    for capability in capabilities:
        for suite_id in registry_capabilities[capability].get("mandatorySuites", []):
            suite = suite_results.get(suite_id)
            if suite is None or suite.get("result") != "pass":
                return invalid("CLAIM_SUITE_INCOMPLETE")

    for evidence in claim["evidence"]:
        if not _is_object(evidence) or evidence.get("subjectBuildId") != claim["host"]["buildId"]:
            return invalid("CLAIM_EVIDENCE_BUILD_MISMATCH")

    for index in range(len(claim["evidence"])):
        state = evidence_states[index] if index < len(evidence_states) else "verifiable"
        if state == "unverifiable":
            return invalid("CLAIM_EVIDENCE_UNVERIFIABLE")

    return invalid("CLAIM_OK")


def execute(input_value):
    """Fixture dispatch facade: operation selects the semantic entry point."""
    input_value = input_value if _is_object(input_value) else {}
    operation = input_value.get("operation")
    if operation == "canonicalize":
        return {"canonical": canonicalize(input_value.get("value"))}
    if operation == "claimDigest":
        return {"digest": claim_digest(input_value.get("claim"))}
    if operation == "validateRegistry":
        return validate_registry(input_value.get("registry"))
    if operation == "validateClaim":
        return validate_claim(input_value.get("claim"), input_value.get("options"))
    raise ValueError("Unknown operation: %s" % operation)
